### DISCUSSION TITLE
How to create a constraint based on the number of objects in an array with a specific value of one of the properties?

### DISCUSSION THREAD

**Discussion by acanqs** [json-schema-org/community#706] [Q&A] [answered]
I'm trying to write a schema that will only pass JSON that has one and only one object of a certain type in an array, regardless of the types of the other objects in that array.

I'll appreciate any help.

The "root" object contains an array of objects. Objects in an array have a `type` property with two possible values "answer" and "comment".
 
 On the list should be:
- one and only one object with the "answer" type
- any number of objects with the "comment" type

The position (index) of the "answer" object on the list is arbitrary.

Valid JSONs
1. There is only one item of the array, and it's "answer":
```
{
  "question": "Foo baz bar?",
  "comments": [
    {
      "type": "answer",
      "content": "Lorem ipsum"
    }
  ]
}
```
2. There are more items in the array, but only one is the "answer":
```
{
  "question": "Foo baz bar?",
  "comments": [
    {
      "type": "comment",
      "content": "Lorem ipsum"
    },
    {
      "type": "comment",
      "content": "Dolor sit amet"
    },
    {
      "type": "answer",
      "content": "Consequentitur flixit"
    }
  ]
}
```

Invalid JSONs:
1. Empty (array or object)
2. There is only one element in the array, but the object type is "comment"
```
{
  "question": "Foo baz bar?",
  "comments": [
    {
      "type": "comment",
      "content": "Lorem ipsum"
    }
  ]
}
```
3. There are more than one element with the "answer" type:
```
{
  "question": "Foo baz bar?",
  "comments": [
    {
      "type": "answer",
      "content": "Lorem ipsum"
    },
    {
      "type": "comment",
      "content": "Dolor sit amet"
    },
    {
      "type": "answer",
      "content": "Consequentitur flixit"
    }
  ]
}
```

I wrote a schema like this:
```
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://somwhere/my-schema.json",
  "title": "Question",
  "type": "object",
  "properties": {
    "question": {
      "type": "string"
    },
    "comments": {
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "type": {
            "enum": ["answer", "comment"]
          },
          "content": {
            "type": "string"
          }
        },
        "required": [ "type", "content" ],
        "additionalProperties": false
      },
      "contains": {
        "type": "object",
        "properties": {
          "type": {
            "enum": ["answer"]
          }
        }
      },
      "minItems": 1,
      "minContains": 1,
      "maxContains": 1
    }
  },
  "required": [ "question", "comments" ],
  "additionalProperties": false
}
```

For validation I use "JSON Schema Validation CLI" (version 3.2.0).
"Valid JSONs" are considered valid by the `jsonschema`. "Invalid JSNOs" 1 and 2 - as expected - invalid. But the third one - passes verification, and I don't know how to build a constraint that will also reject the third case of "Invalid JSONs".

Thanks in advance.

**Comment by gregsdennis** [comment, marked as answer]
Your valids and invalids all check out against that schema on my https://json-everything.net.  Your schema looks right to me.  I'd file a bug on the validator you're using.

**Reply by acanqs** [reply]
Thank you for checking. I can confirm that it works as expected in the "json-everything" validator.
I have a validator from a package on Ubuntu (python3-jsonschema) - and it looks a bit outdated compared to the current version in the source repo (https://github.com/Julian/jsonschema). I will try the latest version.
